File: wingfoil-python/examples/zmq_pub.py
#!/usr/bin/env python3
"""ZMQ publisher example: publishes a counter as bytes every 500ms."""

import struct
import wingfoil as wf

PORT = 5555
print(f"Publishing on tcp://127.0.0.1:{PORT} ...")

(
    wf.ticker(0.5)
    .count()
    .inspect(lambda n: print(f"publishing: {n}"))
    .map(lambda n: struct.pack(">Q", n))  # u64 big-endian bytes
    .zmq_pub(PORT)
    .run(realtime=True)
)

File: wingfoil/examples/zmq/direct/zmq_pub.rs
// ZMQ publisher — direct mode (no service discovery).
//
// Binds a PUB socket on pub_port and publishes a counter every 100 ms.
// The subscriber connects to the address directly.
//
// Run publisher and subscriber in separate terminals:
//
//   RUST_LOG=info cargo run --example zmq_direct_pub --features zmq-beta
//   RUST_LOG=info cargo run --example zmq_direct_sub --features zmq-beta

use log::Level::Info;
use std::time::Duration;
use wingfoil::adapters::zmq::ZeroMqPub;
use wingfoil::*;

fn main() -> anyhow::Result<()> {
    env_logger::init();

    let pub_port = 7779u16;

    ticker(Duration::from_millis(100))
        .count()
        .logged("pub", Info)
        .zmq_pub(pub_port, ())
        .run(RunMode::RealTime, RunFor::Forever)?;

    Ok(())
}
